### PR TITLE
use better initial guess

### PR DESCRIPTION
**Pull request by Yaffle**
should make `sqrt(2n * 10n**10000n)` much faster, otherwise it has linear convergence during first iterations

### CODE REVIEW

**Comment by wolfram77** [comment]
Some interesting numbers to sqrt ...

I was reading your update for `sqrt()`. `x.toString(16).length` is equivalent to `δ + log16(x)`, where `δ = (0, 1]`. So the `initialGuess = 1n << BigInt(x.toString(16).length * 2)` is equivalent to `2 ^ (2δ + 2log16(x))`, which simplifies to `4^δ * √x`. Therefore, `initialGuess` is always larger than `√x`. It is however possible to tighten this bound (the `4^δ` term) by using `2 ^ (δ + log4(x))` to `2^δ`. Or even better with `2 ^ (δ/2 + log2(x)/2)` to `2^(δ/2)`. For this, `initialGuess` would be `1n << BigInt(x.toString(2).length / 2)`.

**Comment by Yaffle** [comment]
@wolfram77 , you are right, but x.toString(16) is much faster then x.toString(2), of course, you can still use your log2(x) function, and reimplement it to use x.toString(16) and analyze the first digit for efficiency
I hope, log2(x) will be available natively.

**Comment by Yaffle** [comment]
@wolfram77 , I have changed the change to use imported log2

**Comment by wolfram77** [comment]
Is `x.toString(16)` faster because it has to generate a smaller string? Even if log2(x) isnt available natively, devs might optimize for `x.toString(2).length`.

**Comment by Yaffle** [comment]
@wolfram77 , may be, but today x.toString(16) is faster
there is a proposal to add log2 for bigints - https://github.com/js-choi/proposal-bigint-math

File: src/root.ts [hunk 1]
```diff
@@ -1,5 +1,11 @@
+import log2 from './log2';
+
 function rootPos(x: bigint, n: bigint): bigint {
-  var a = x, b = x+1n, m = n-1n;
+  const initialGuess = 1n << ((log2(x) / n) + 1n);
+  if (initialGuess === 2n) {
```
Comment (wolfram77) [inline src/root.ts:5]: Is there an edge case here?

File: src/root.ts [hunk 2]
```diff
@@ -1,5 +1,11 @@
+import log2 from './log2';
+
 function rootPos(x: bigint, n: bigint): bigint {
-  var a = x, b = x+1n, m = n-1n;
+  const initialGuess = 1n << ((log2(x) / n) + 1n);
```
Comment (wolfram77) [inline src/root.ts:4]: This can be `1n << ((1n + log2(x)) / n)`?

File: src/sqrt.ts [hunk 3]
```diff
@@ -1,5 +1,8 @@
+import log2 from './log2';
+
 function sqrtPos(x: bigint): bigint {
-  var a = x, b = x+1n;
+  const initialGuess = 1n << (log2(x) / 2n + 1n);
```
Comment (wolfram77) [inline src/sqrt.ts:4]: This can be `1n << ((1n + log2(x)) / 2n)`?

**Review by wolfram77 — CHANGES_REQUESTED** [review]

**Comment by Yaffle** [comment]
@wolfram77 , btw, for even better initial approximation it is needed to recursively call itself with half of the binary digits:
```
sqrt(x >> (quarter * 2n)) << quarter
```

File: src/sqrt.ts (hunk 3 above)
Comment (Yaffle) [reply src/sqrt.ts:4]: @wolfram77 , not sure, this could make it less than mathematical value

**Comment by wolfram77** [comment]
Woah, i didnt understand the recursive, can you explain it?

File: src/root.ts (hunk 1 above)
Comment (Yaffle) [reply src/root.ts:5]: my tests fail with "RangeError: Maximum BigInt size exceeded" without this branch

**Comment by Yaffle** [comment]
@wolfram77 
```javascript
const qurter = log2(x) / 4n;
const initialGuess = sqrt(x >> (quarter * 2n)) << quarter; // well, we need to make  it large then the mathematical value by + (1n << quarter)
```
your are doing recursive call to get better intial guess (where almost half of the binary digits are correct), so it is only needed to run the Newton's iteration 1-2 times.

**Comment by wolfram77** [comment]
So if the function `log2(x) {}` is equivalent to `log2(x) - δ`, where `δ = [0, 1)`, then `quarter` simplifies to `log2(x)/4 - δ/4`. When i tried to simplify `initialGuess`, i get `sqrt(√(2^δ) * √x) * ∜x/∜(2^δ)`. So the `2^δ` part cancels out and might be `√x` remains. Is this correct? Great! Is this in a paper? One issue left is to fix recursive termination condition. Maybe i should merge the previous change and try this one next.

**Comment by Yaffle** [comment]
@wolfram77 , thanks for merging!

**Pull request merged**